openverse_catalog/dags/providers/provider_api_scripts/time_delineated_provider_data_ingester.py

```diff
@@ -0,0 +1,238 @@
+import logging
+from abc import abstractmethod
+from datetime import datetime, timedelta, timezone
+
+from airflow.exceptions import AirflowException
+from providers.provider_api_scripts.provider_data_ingester import ProviderDataIngester
+
+
+logger = logging.getLogger(__name__)
+
+
+class TimeDelineatedProviderDataIngester(ProviderDataIngester):
+    """
+    A subclass of ProviderDataIngester which breaks the ingestion interval into
+    smaller intervals, and runs ingestion separately for each interval.
+
+    Must be used on a dated DAG.
+
+    Class variables of note:
+    max_records:        integer giving the maximum number of records to be expected in
+                        each interval
+    division_threshold: integer threshold number of records in an hour, used to
+                        determine how many divisions to use
+    min_divisions:      integer giving the number of divisions to break an hour into if
+                        the division_threshold is not exceeded
+    max_divisions:      integer giving the number of divisions to break an hour into if
+                        the division_threshold is exceeded
+    should_raise_error: whether to raise an exception when more records are retrieved
+                        from the API than the API reports to expect
+    """
+
+    min_divisions = 4  # 15-minute default intervals
+    max_divisions = 12  # 5-minute default intervals
+    should_raise_error = True
+
+    @property
+    @abstractmethod
+    def max_records(self) -> int:
+        pass
+
+    @property
+    @abstractmethod
+    def division_threshold(self) -> int:
+        pass
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+
+        # This class is used to generate timestamp intervals over a 24-hour period,
+        # so it should only be used for a dated DAG.
+        assert (
+            self.date is not None
+        ), f"{self.__class__.__name__} should only be used for dated DAGs."
```

Comment:
In general I think we'd like to avoid assertions in code that is run in production (the main reason for this being that [assertions can be disabled entirely at runtime and thus ignored](https://stackoverflow.com/questions/1273211/disable-assertions-in-python)). Do you mind changing this to raise a `ValueError` instead?
```suggestion
        # This class is used to generate timestamp intervals over a 24-hour period,
        # so it should only be used for a dated DAG.
        if self.date is None:
            raise ValueError(
                f"{self.__class__.__name__} should only be used for dated DAGs."
            )
```

Reply:
You're absolutely right, good call. Done ✅